fv3gfs/util/communicator.py

```diff
@@ -310,7 +315,7 @@ def tile(self) -> TileCommunicator:
         """communicator for within a tile"""
         if self._tile_communicator is None:
             self._initialize_tile_communicator()
-        return self._tile_communicator
+        return cast(TileCommunicator, self._tile_communicator)
```

Comment:
Is the "lazy" loading really necessary for performance reasons or as a means to accomplish polymorphism? The behavior seems too dynamic to really benefit from a static type-checker. If there are contexts where the `tile` property is never actually accessed (e.g. on non-root workers), a more static design based on inheritance or compisition would be a lot more friendly on they type-checker.

Reply:
I'm repeating myself here, but the public API is static and based on composition. The "optional" type is private and only used in one place (the lazy loader). But the implication here is that lazy loading is incompatible with static typing, which isn't the case. See the first example [here](https://en.wikipedia.org/wiki/Lazy_loading) of statically-typed lazy loading in C#. I always feel weird about Python's treatment of None vs static languages' treatment of null in these cases.

I'm not sure whether or not the lazy loading is necessary, but if you are using the communicator only for halo updates and not for scatter/gather on a tile the tile attribute will likely not be accessed.

Reply:
Okay. It sounds like it might not be worth creating a different class to isolate the `.tile` attribute. 

>But the implication here is that lazy loading is incompatible with static typing, which isn't the case.

I wasn't saying that. The if-statement I suggested above implements a statically-typed approach to lazy loading without an explicit "cast". I only meant that polymorphism via lazy-loading would be nice to avoid if you can easily predict how an object will be used at initialization. It doesn't sound like that is the case here, so lazy-loading is probably fine.

Reply:
Also, I'd be interested to see what happens if implemented exactly like the C# example:
```
# use a function with a return
self._tile_communicator = self._initialize_tile_communicator()
return self._tile_communicator
```
It's possible mypy would be smart enough to know that `._tile_communicator` is not None in this instance.

Reply:
Ah yeah it would probably work. In other words, we can remove the `cast()` if `self._initialize_tile_communicator` returns its value explicitly instead of setting it behind the scenes.

I'm good with that change for a future PR.